quickwit-serve/src/index_api/rest_handler.rs

```diff
@@ -0,0 +1,305 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::convert::Infallible;
+use std::sync::Arc;
+
+use quickwit_config::IndexConfig;
+use quickwit_core::IndexService;
+use quickwit_search::SearchError;
+use serde::de::DeserializeOwned;
+use tracing::info;
+use warp::{Filter, Rejection};
+
+use crate::format::Format;
+
+pub fn index_management_handlers(
+    index_service: Arc<IndexService>,
+) -> impl Filter<Extract = impl warp::Reply, Error = Rejection> + Clone {
+    get_index_metadata_handler(index_service.clone())
+        .or(get_indexes_metadatas_handler(index_service.clone()))
+        .or(create_index_handler(index_service.clone()))
+        .or(delete_index_handler(index_service))
+}
+
+fn get_index_metadata_handler(
+    index_service: Arc<IndexService>,
+) -> impl Filter<Extract = impl warp::Reply, Error = Rejection> + Clone {
+    warp::path!("indexes" / String)
+        .and(warp::get())
+        .and(warp::any().map(move || index_service.clone()))
+        .and_then(get_index_metadata)
+}
+
+async fn get_index_metadata(
+    index_id: String,
+    index_service: Arc<IndexService>,
+) -> Result<impl warp::Reply, Infallible> {
+    info!(index_id = %index_id, "get-index");
+    let index_metadata = index_service
+        .get_index(&index_id)
+        .await
+        .map_err(SearchError::from);
```

Comment:
Maybe we could have `IndexError` struct that just implement 
```rust 
struct IndexError;
impl warp::reject::Reject for IndexError {}
```
Maybe we could also rename SearchError into something more generic since we now have more services.
I suggest we create a small issue for this rather than you handling it here.

Reply:
Good point.

Reply:
I created an custom `IndexServiceError`. I used the same pattern as the one used for other service errors.